cli/src/cmd_instance_serial.rs

```diff
@@ -0,0 +1,216 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2023 Oxide Computer Company
+
+use anyhow::Result;
+use clap::Parser;
+use oxide_api::types::NameOrId;
+use oxide_api::ClientInstancesExt;
+use std::path::PathBuf;
+
+#[derive(Parser, Debug, Clone)]
+#[clap(verbatim_doc_comment)]
+#[clap(name = "instance")]
+pub enum CmdInstance {
+    Serial(CmdInstanceSerial),
+}
+
+impl CmdInstance {
+    pub async fn run(&self, ctx: &mut crate::context::Context) -> Result<()> {
+        match &self {
+            Self::Serial(cmd) => cmd.run(ctx).await,
+        }
+    }
+}
+
+/// Connect to or retrieve data from the instance's serial console.
+#[derive(Parser, Debug, Clone)]
+#[clap(verbatim_doc_comment)]
+#[clap(name = "serial")]
+pub struct CmdInstanceSerial {
+    #[clap(subcommand)]
+    subcmd: SubCommand,
+}
+
+impl CmdInstanceSerial {
+    pub async fn run(&self, ctx: &mut crate::context::Context) -> Result<()> {
+        match &self.subcmd {
+            SubCommand::Console(cmd) => cmd.run(ctx).await,
+            SubCommand::History(cmd) => cmd.run(ctx).await,
+        }
+    }
+}
+
+#[derive(Parser, Debug, Clone)]
+enum SubCommand {
+    Console(CmdInstanceSerialConsole),
+    History(CmdInstanceSerialHistory),
+}
+
+/// Connect to an instance's serial console interactively.
+#[derive(Parser, Debug, Clone)]
+#[clap(verbatim_doc_comment)]
+#[clap(name = "console")]
+pub struct CmdInstanceSerialConsole {
+    /// Name or ID of the instance
+    #[clap(long, short)]
+    instance: NameOrId,
+
+    /// Name or ID of the project
+    #[clap(long, short)]
+    project: Option<NameOrId>,
```

Comment:
Added a specific error message for this in the CLI since it's something we *can* catch before the request goes out in this case, but this also is a symptom of a more general issue where errors occurring before the WS proxy event loop in nexus was started wouldn't be communicated to the client at all:
- surfacing these in nexus https://github.com/oxidecomputer/omicron/pull/2871/commits/1456c75342234d17e50f2c909ddc3c40dfa07ce4
- client-side fix (used by this) https://github.com/oxidecomputer/thouart/pull/2